lib/storage: add tracker for time series metric names usage (#8118)
VictoriaMetrics/VictoriaMetrics: docs/README.md

This may lead to inflated values when samples for the same time series are spread across multiple vmstorage nodes
due to [replication](#replication) or [rerouting](https://docs.victoriametrics.com/cluster-victoriametrics/?highlight=re-routes#cluster-availability).

### Track ingested metrics usage

`VictoriaMetrics` provides an ability to explore time series metric names used at Query Requests. And exposes an API endpoint - `/api/v1/status/metric_names_usage_stats`.
This endpoints requires flag `--storage.trackMetricNamesUsage` to be set. It accepts the following query parameters:

* `limit` - limits an output records count.

> * `lte` - less than equal number for query requests count. By default, it's empty and returns all records. But it could be set to `0` and return only metrics without registered query requests.

Reviewer: `le` - less than or equal to ...

* `pattern` - a substring pattern to match metric names.

 The API endpoint returns the following `JSON` response:

```json
{
  "status": "success",
  "stats_collected_since": 1737534094,
  "stats_collected_records_total": 2,
  "records": [
    {
      "metric_name": "node_disk_writes_completed_total",
      "query_requests": 0,
      "last_request_timestamp": 1737534262
    },
    {
      "metric_name": "node_network_transmit_errs_total",
      "query_requests": 100,
      "last_request_timestamp": 1737534262
    }
  ]
}
```

>  `VictoriaMetrics` stores tracked metric names in-memory and saves state `on-disk` at `data/cache` folder. State could be reset via API call `/api/v1/admin/tsdb/reset_metric_names_usage` or per storage via [cache removal](#cacheremoval).

Reviewer: s/ at / in /

>  `VictoriaMetrics` stores tracked metric names in-memory and saves state `on-disk` at `data/cache` folder. State could be reset via API call `/api/v1/admin/tsdb/reset_metric_names_usage` or by manually per storage with [cache removal](#Cacheremoval).

Reviewer 2: ```suggestion
 `VictoriaMetrics` stores tracked metric names in-memory and saves state `on-disk` at `data/cache` folder. State could be reset via API call `/api/v1/admin/tsdb/reset_metric_names_usage` or per storage via [cache removal](#cacheremoval).
```

This may lead to inflated values when samples for the same time series are spread across multiple vmstorage nodes
due to [replication](#replication) or [rerouting](https://docs.victoriametrics.com/cluster-victoriametrics/?highlight=re-routes#cluster-availability).

### Track ingested metrics usage statistics

> `VictoriaMetrics` provides an ability to explore time series metric names used at Query Requests. This feature can be enabled with flag `--storage.trackMetricNamesStats`.

Reviewer: `... used in query requests` (no need to capitalize, these are just regular words here)

> Statistics exposed at API endpoint - `/prometheus/api/v1/status/metric_names_stats`. It accepts the following query parameters:

Reviewer: Statistics is exposed at  `/api/v1/status/metric_names_stats` API endpoint

Also, is `/prometheus` a part of the path?

* `limit` - limits an output records count.

> * `le` - A "less than or equal" threshold for the query request count. By default, this is empty, returning all records. If set to 0, it returns only metrics without any registered query requests.

Reviewer: `less than or equal` -> `less than or equal to`

* `match_pattern` - a substring pattern to match metric names.

 The API endpoint returns the following `JSON` response:

```json
{
  "status": "success",
  "statsSollectedSince": 1737534094,
  "statsCollectedRecordsTotal": 2,
  "records": [
    {
      "metricName": "node_disk_writes_completed_total",
      "queryRequests": 0,
      "lastRequestTimestamp": 1737534262
    },
    {
      "metricName": "node_network_transmit_errs_total",
      "queryRequestsCount": 100,
      "lastRequestTimestamp": 1737534262
    }
  ]
}
```

 VictoriaMetrics stores tracked metric names in memory and saves the state to disk in the data/cache folder. 
The size of the in-memory state is limited to 1% of the available memory by default.
This limit can be adjusted using the -storage.cacheSizeMetricNamesStats flag.

>  When the maximum state capacity is reached, VictoriaMetrics will reject newly registered time series.

Reviewer: `VictoriaMetrics will reject newly registered time series` - this could be understood that VictoriaMetrics will reject the actual samples during the ingestion. Please rewrite so it is not ambiguos.

However, read request statistics for already tracked time series will continue to work as expected.

> V ictoriaMetrics exposes the following metrics for the tracker state:

Reviewer: `V ictoriaMetrics` -> `VictoriaMetrics`

This may lead to inflated values when samples for the same time series are spread across multiple vmstorage nodes
due to [replication](#replication) or [rerouting](https://docs.victoriametrics.com/cluster-victoriametrics/?highlight=re-routes#cluster-availability).

### Track ingested metrics usage

VictoriaMetrics provides an ability to record statistics of fetched [metric names](https://docs.victoriametrics.com/keyconcepts/#structure-of-a-metric) during [querying](https://docs.victoriametrics.com/keyconcepts/#query-data). This feature can be enabled via flag `--storage.trackMetricNamesStats` (disabled by default) on  single-node VictoriaMetrics or [vmstorage](https://docs.victoriametrics.com/cluster-victoriametrics/#architecture-overview).

To get metric names usage statistic use  `/prometheus/api/v1/status/metric_names_stats` API endpoint. It accepts the following query parameters:

* `limit` - limits the number of metric names in response. By default, API returns 1000 records.

> * `le` -  "less than or equal" threshold for filtering metric names by their usage count in queries. For example, if set to 1 then API returns only metric names that were queried 1 or less times.

Reviewer: `less than or equal` -> `less than or equal to`

* `match_pattern` - a substring pattern to match metric names.

 The API endpoint returns the following `JSON` response:

```json
{
  "status": "success",
  "statsSollectedSince": 1737534094,
  "statsCollectedRecordsTotal": 2,
  "records": [
    {
      "metricName": "node_disk_writes_completed_total",
      "queryRequests": 0,
      "lastRequestTimestamp": 1737534262
    },
    {
      "metricName": "node_network_transmit_errs_total",
      "queryRequestsCount": 100,
      "lastRequestTimestamp": 1737534262
    }
  ]
}
```

VictoriaMetrics stores tracked metric names in memory and saves the state to disk in the data/cache folder during restarts.
The size of the in-memory state is limited to 1% of the available memory by default.
This limit can be adjusted using the `-storage.cacheSizeMetricNamesStats` flag.

> When the maximum state capacity is reached, VictoriaMetrics will not track stats newly registered time series.

Reviewer: for newly

This may lead to inflated values when samples for the same time series are spread across multiple vmstorage nodes
due to [replication](#replication) or [rerouting](https://docs.victoriametrics.com/cluster-victoriametrics/?highlight=re-routes#cluster-availability).

> ### Track ingested metrics usage

Reviewer 2: It doesn't mention which HTTP API calls contirbute to the stats. Could you please add this to the docs?

Author: It's good point actually. Currently, it will track requests for export ( if reduce_mem_usage=false) and federate API calls. It'll only mislead users. 

 Unfortunately, it requires to add additional option to search request, which will affect vmselect-> vmstorage communication. new version of vmselect cannot be used with older version of vmstorage. However, old version of vmselect can be used with new version of vmstorage. It must provide a zero downtime migration path - at first update storage nodes, after that update select nodes.

> VictoriaMetrics provides an ability to record statistics of fetched [metric names](https://docs.victoriametrics.com/keyconcepts/#structure-of-a-metric) during [querying](https://docs.victoriametrics.com/keyconcepts/#query-data). This feature can be enabled via flag `--storage.trackMetricNamesStats` (disabled by default) on  single-node VictoriaMetrics or [vmstorage](https://docs.victoriametrics.com/cluster-victoriametrics/#architecture-overview). Querying metric with non-matching filters doesn't increase counter.

Reviewer 2: ```suggestion
VictoriaMetrics provides the ability to record statistics of fetched [metric names](https://docs.victoriametrics.com/keyconcepts/#structure-of-a-metric) during [querying](https://docs.victoriametrics.com/keyconcepts/#query-data). This feature can be enabled via the flag `--storage.trackMetricNamesStats` (disabled by default) on  a single-node VictoriaMetrics or [vmstorage](https://docs.victoriametrics.com/cluster-victoriametrics/#architecture-overview). Querying a metric with non-matching filters doesn't increase the counter for this particular metric name.
```

> For example, querying for `vm_log_messages_total{level!="info"}` won't increment `vm_log_messages_total` counter because there are no `level="error"` metrics yet.

Reviewer 2: ```suggestion
For example, querying for `vm_log_messages_total{level!="info"}` won't increment usage counter for `vm_log_messages_total` if there are no `{level="error"}` or `{level="warning"}` series yet.
```

> To get metric names usage statistic use  `/prometheus/api/v1/status/metric_names_stats` API endpoint. It accepts the following query parameters:

Reviewer 2: ```suggestion
To get metric names usage statistics, use the `/prometheus/api/v1/status/metric_names_stats` API endpoint. It accepts the following query parameters:
```

> * `limit` - limits the number of metric names in response. By default, API returns 1000 records.

Reviewer 2: ```suggestion
* `limit` - integer value to limit the number of metric names in response. By default, API returns 1000 records.
```

> * `le` -  less than or equal to threshold for filtering metric names by their usage count in queries. For example, if set to 1 then API returns only metric names that were queried 1 or less times.

Reviewer 2: ```suggestion
* `le` -  `less than or equal`, is an integer threshold for filtering metric names by their usage count in queries. For example, with `?le=1` API returns metric names that were queried <=1 times.
```

> * `match_pattern` - a substring pattern to match metric names. For example, `match_pattern=vm_` will match any metric names with `vm_` pattern, like `vm_http_requests`, `max_vm_memory_available`. It doesn't support regex syntax.

Reviewer 2: ```suggestion
* `match_pattern` - a substring pattern to match metric names. For example, `?match_pattern=vm_` will match any metric names with `vm_` pattern, like `vm_http_requests`, `max_vm_memory_available`. It doesn't support regex syntax.
```

 The API endpoint returns the following `JSON` response:

```json
{
  "status": "success",
  "statsSollectedSince": 1737534094,
  "statsCollectedRecordsTotal": 2,
  "records": [
    {
      "metricName": "node_disk_writes_completed_total",
      "queryRequests": 0,

>       "lastRequestTimestamp": 1737534262

Reviewer 2: how it could have lastRequestTimestamp>0 and queryRequests==0 in the same time?

    },
    {
      "metricName": "node_network_transmit_errs_total",
      "queryRequestsCount": 100,
      "lastRequestTimestamp": 1737534262
    }
  ]
}
```

VictoriaMetrics stores tracked metric names in memory and saves the state to disk in the data/cache folder during restarts.
The size of the in-memory state is limited to 1% of the available memory by default.
This limit can be adjusted using the `-storage.cacheSizeMetricNamesStats` flag.

> When the maximum state capacity is reached, VictoriaMetrics will not track stats for newly registered time series.

Reviewer 2: ```suggestion
When the maximum state capacity is reached, VictoriaMetrics will stop tracking stats for newly registered time series.
```

However, read request statistics for already tracked time series will continue to work as expected.

VictoriaMetrics exposes the following metrics for the metric name tracker:
* vm_cache_size_bytes{type="storage/metricNamesStatsTracker"}
* vm_cache_size{type="storage/metricNamesStatsTracker"}
* vm_cache_size_max_bytes{type="storage/metricNamesStatsTracker"}

> The metric name tracker state can be reset via the API endpoint /api/v1/admin/status/metric_names_stats/reset or  

Reviewer 2: ```suggestion

An alerting rule with query `vm_cache_size_bytes{type="storage/metricNamesStatsTracker"} \ vm_cache_size_max_bytes{type="storage/metricNamesStatsTracker"} > 0.9` can be used to notify the user of cache utilization exceeding 90%.

The metric name tracker state can be reset via the API endpoint /api/v1/admin/status/metric_names_stats/reset or  
```

  -reloadAuthKey value
     Auth key for /-/reload http endpoint. It must be passed via authKey query arg. It overrides -httpAuth.*
     Flag value can be read from the given file when using -reloadAuthKey=file:///abs/path/to/file or -reloadAuthKey=file://./relative/path/to/file . Flag value can be read from the given http/https url when using -reloadAuthKey=http://host/path or -reloadAuthKey=https://host/path
  -resetMetricsUsageAuthKey value

>         authKey for reseting metric names usage cache via /api/v1/admin/tsdb/reset_metric_names_usage. It overrides -httpAuth.*

Reviewer: Please fix the the API path
